src/NServiceBus.Transport.SQS/MessageDispatcher.cs

```diff
@@ -341,24 +337,61 @@ async Task<TMessage> PrepareMessage<TMessage>(IOutgoingTransportOperation transp
             return preparedMessage;
         }
 
-        async Task ApplyMulticastOperationMappingIfNecessary(MulticastTransportOperation transportOperation, SnsPreparedMessage snsPreparedMessage, CancellationToken cancellationToken)
+        async Task<SnsPreparedMessage> PrepareMessage(MulticastTransportOperation transportOperation, CancellationToken cancellationToken)
         {
-            if (transportOperation == null || snsPreparedMessage == null)
+            var sqsTransportMessage = new TransportMessage(transportOperation.Message, transportOperation.Properties, encodeBodyToBase64);
+
+            var messageId = transportOperation.Message.MessageId;
+
+            var preparedMessage = new SnsPreparedMessage();
+
+            await ApplyMulticastOperationMapping(transportOperation, preparedMessage, cancellationToken).ConfigureAwait(false);
+
+            preparedMessage.Body = JsonSerializer.Serialize(sqsTransportMessage, transportMessageSerializerOptions);
+            preparedMessage.MessageId = messageId;
+
+            preparedMessage.CalculateSize();
+            if (preparedMessage.Size <= TransportConstraints.MaximumMessageSize)
             {
-                return;
+                return preparedMessage;
+            }
+
+            if (s3 == null)
+            {
+                throw new Exception("Cannot send large message because no S3 bucket was configured. Add an S3 bucket name to your configuration.");
+            }
+
+            var key = $"{s3.KeyPrefix}/{messageId}";
+            using (var bodyStream = new ReadonlyStream(transportOperation.Message.Body))
+            {
+                var putObjectRequest = new PutObjectRequest
+                {
+                    BucketName = s3.BucketName,
+                    InputStream = bodyStream,
+                    Key = key
+                };
+
+                s3.NullSafeEncryption.ModifyPutRequest(putObjectRequest);
+
+                await s3.S3Client.PutObjectAsync(putObjectRequest, cancellationToken).ConfigureAwait(false);
             }
```

Comment:
We should be able to clean up some of this repetition when we add in the `doNotWrapOutgoingMessages` changes. Certainly the pushing to s3 bit at least.

Reply:
Agreed